src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -1261,19 +1544,682 @@ private bool ReadInput()
+                {
+                    currentCharacterTokenRelativeIndex++;
+                }
+                else
+                {
+                    break;
+                }
+            }
+            while ((this.tokenStartIndex + currentCharacterTokenRelativeIndex) < this.storedCharacterCount || await this.ReadInputAsync().ConfigureAwait(false));
+
+            return this.ConsumeTokenToString(currentCharacterTokenRelativeIndex);
+        }
+
+        /// <summary>
+        /// Asynchronously reads bytes from the current string value.
+        /// </summary>
+        /// <param name="chars">The character buffer to populate</param>
+        /// <param name="offset">The number of characters offset into the buffer to read</param>
+        /// <param name="maxLength">The maximum number of characters to read into the buffer</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains the number of characters read.</returns>
+        /// <remarks>
+        /// Assumes that the current token position points to the opening quote.
+        /// Note that the string parsing can never end with EndOfInput, since we're already seen the quote.
+        /// So it can either return a string successfully or fail.</remarks>
+        [SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity", Justification = "Splitting the function would make it hard to understand.")]
+        private async Task<int> ReadCharsAsync(char[] chars, int offset, int maxLength)
+        {
+            if (!readingStream)
+            {
+                return 0;
+            }
+
+            int charsRead = 0;
+
+            while (charsRead < maxLength && (this.tokenStartIndex < this.storedCharacterCount || await this.ReadInputAsync().ConfigureAwait(false)))
+            {
+                char character = this.characterBuffer[this.tokenStartIndex];
+                bool advance = true;
+
+                if (character == GetClosingQuoteCharacter(this.streamOpeningQuoteCharacter) && --this.balancedQuoteCount < 1)
+                {
+                    if (character != '"')
+                    {
+                        // The character is part of the JSON stream; copy it to the output buffer
+                        chars[charsRead + offset] = character;
+                        charsRead++;
+                        this.scopes.Pop();
+                    }
+
+                    // Consume the closing quote character.
+                    this.tokenStartIndex++;
+                    readingStream = false;
+                    this.scopes.Peek().ValueCount++;
+
+                    // move to next node
+                    await this.ReadAsync().ConfigureAwait(false);
+                    return charsRead;
+                }
+
+                if (character == this.streamOpeningQuoteCharacter)
+                {
+                    this.balancedQuoteCount++;

```

Comment:
I think that this is just trying to undo the decrement from line 2014. If that's the case, it would probably make more sense to just have line 2014 read `... && this.balancedQuoteCount - 1 < 1`

Reply:
@corranrogue9 Note that on line 2014 there's the condition that has to be satisfied before the decrement happens is `character == GetClosingQuoteCharacter(this.streamOpeningQuoteCharacter)` which is a little different from the condition for incrementing `character == this.streamOpeningQuoteCharacter`